### PR TITLE
Skip greylisting when we can't talk to greylist DB

### PR DESCRIPTION
**Pull request by jaredj**
This is split into two commits to hopefully ease reviewing, but can be squashed before merge if desired.

### CODE REVIEW

File: lib/Qpsmtpd/Plugin.pm [hunk 1]
```diff
@@ -347,10 +347,16 @@ sub _register_standard_hooks {
     }
 }
 
+sub db_args {
+    my ( $self, %arg ) = @_;
```
Comment (msimerson) [inline lib/Qpsmtpd/Plugin.pm]: this seems likely to encounter problems when the wrong number of arguments is passed in (due to an error elsewhere). Validating the arguments (at least the correct qty) seems prudent.

**Comment by msimerson** [comment]
Should this change have some POD accompanying it?

File: lib/Qpsmtpd/Plugin.pm (hunk 1 above)
Comment (jaredj) [inline lib/Qpsmtpd/Plugin.pm]: should we die if the wrong number of args are passed, or just warn?  if all we want is to warn, perl already does this for us:

```
perl -Mstrict -wle 'my %h = (1,2,3)'
Odd number of elements in hash assignment at -e line 1.
```

**Comment by jaredj** [comment]
I can't think of any POD needed... a subsequent PR will probably add a new plugin arg for optionally deferring mail while e.g. redis is down, which will need POD.

File: lib/Qpsmtpd/Plugin.pm (hunk 1 above)
Comment (msimerson) [inline lib/Qpsmtpd/Plugin.pm]: * We know that that errors in the logs can go unnoticed for a long time
* it's a programming error we want raised early, preferably early enough that it's caught in testing.
* we don't want a plugin failure to take down the mail server, but IIRC, plugin exceptions are handled.

**Comment by msimerson** [comment]
> I can't think of any POD needed.

What happens when QP starts up and Redis isn't available? How does greylisting recover? Does it attempt connecting upon every connection? How long does that delay every connection when Redis isn't available?

File: lib/Qpsmtpd/Plugin.pm (hunk 1 above)
Comment (msimerson) [inline lib/Qpsmtpd/Plugin.pm]: > perl already does this for us <snip> Odd number of elements in hash assignment at -e line 1.

That error message is considerably worse than, "invalid arguments in call to db_vars, cowardly refusing to answer."

**Comment by jaredj** [comment]
On every connection the greylist plugin will attempt to connect to redis, and if it can't connect it will bail. AFAICT bailing won't really present any notable delay.

Are you thinking the answers to these questions should appear in some POD somewhere?

File: lib/Qpsmtpd/Plugin.pm (hunk 1 above)
Comment (jaredj) [inline lib/Qpsmtpd/Plugin.pm]: makes sense. working on this.

**Comment by msimerson** [comment]
Most databases have a connection timeout setting, so I'm assuming Redis does too. I'd like to know what the cost to each connection is in the face of a Redis outage, and I assume that that information might be useful to you too. Maybe it's not entirely necessary to document it in POD, but knowing it and having it documented **somewhere** would help us decide if it's sensible to "remember" DB failures and not attempt reconnecting more often than once per N seconds. This could make a very large difference in performance for busy mail servers. If QP::DB ever supported other DB engines, this could get mandatory as those servers will block clients that attempt connecting *too* often.

**Comment by jaredj** [comment]
I was thinking more along the lines of the redis server actually being stopped, in which case the failure would be immediate.  Let me investigate the connect timeout. If the default is a bit long, I'd suggest supplying a timeout that's short enough that we can tolerate it without the extra logic to minimize connection attempts.

File: lib/Qpsmtpd/Plugin.pm (hunk 1 above)
Comment (jaredj) [inline lib/Qpsmtpd/Plugin.pm]: 42c5519

**Comment by jaredj** [comment]
The redis timeout is not documented, but it turns out it's somewhere around 75 seconds.  15a2973 sets it to one second and adds a bit of documentation about what happens when redis is down.  This can be squashed further after review if desired

**Pull request merged**